### PR TITLE
ENH: Extending _distance_pybind with additional distance metrics (continued)

### PR DESCRIPTION
**Pull request by jjerphan**
#### Reference issue

Follow-up of #14640, authored by @czgdp1807.

Alternative to #17522 whose CI is failing continuously.

> #### Reference issue
> 
> #14077
>
> #### What does this implement/fix?
> 
> In this PR following metrics will be ported to `_distance_pybind11`,
>  * [x]  dice
>  * [x]  hamming
>  * [x]  jaccard
>  * [x]  kulsinski
>  * [x]  matching
>  * [x]  rogerstanimoto
>  * [x]  russellrao
>  * [x]  sokalmichener
>  * [x]  sokalsneath
>  * [x]  yule
> 
> 
> The remaining distance metrics (tricky and potential blocker ones) are being dealt with in #14611
> 
> Please feel free to leave your reviews.
> #### Additional information
> 
> Benchmark results.
> 
> ```python
>        before           after         ratio
>      [453aad79]       [24897be1]
>      <master>         <pybind11_dist_framework>
> +     1.59±0.02ms      1.98±0.04ms     1.24  spatial.Xdist.time_pdist(1000, 'dice')
> +     3.20±0.01ms      3.89±0.02ms     1.21  spatial.Xdist.time_cdist(1000, 'dice')
> +     31.9±0.07μs       38.5±0.5μs     1.21  spatial.Xdist.time_cdist(100, 'dice')
> +     17.0±0.07μs      20.2±0.06μs     1.19  spatial.Xdist.time_pdist(100, 'dice')
> +     1.90±0.01ms      2.19±0.03ms     1.15  spatial.Xdist.time_pdist(1000, 'yule')
> +     3.85±0.02ms      4.40±0.07ms     1.14  spatial.Xdist.time_cdist(1000, 'yule')
> +     20.2±0.07μs       23.0±0.3μs     1.14  spatial.Xdist.time_pdist(100, 'yule')
> +     38.3±0.02μs       43.2±0.2μs     1.13  spatial.Xdist.time_cdist(100, 'yule')
> +     19.9±0.07μs       22.0±0.3μs     1.11  spatial.Xdist.time_pdist(100, 'jaccard')
> +      38.0±0.2μs      41.6±0.06μs     1.09  spatial.Xdist.time_cdist(100, 'jaccard')
> +     1.35±0.01ms      1.47±0.02ms     1.09  spatial.Xdist.time_pdist(1000, 'russellrao')
> +     1.94±0.01ms      2.10±0.03ms     1.08  spatial.Xdist.time_pdist(1000, 'jaccard')
> +     2.73±0.03ms       2.95±0.1ms     1.08  spatial.Xdist.time_cdist(1000, 'sokalmichener')
> +     3.94±0.03ms       4.23±0.1ms     1.07  spatial.Xdist.time_cdist(1000, 'jaccard')
> +     14.7±0.04μs      15.6±0.04μs     1.06  spatial.Xdist.time_pdist(100, 'sokalmichener')
> -     31.9±0.05μs       30.2±0.1μs     0.95  spatial.Xdist.time_cdist(100, 'sokalsneath')
> -     2.33±0.02μs      2.20±0.01μs     0.94  spatial.Xdist.time_pdist(10, 'jaccard')
> -     3.04±0.01μs      2.84±0.07μs     0.94  spatial.XdistWeighted.time_pdist(20, 'euclidean')
> -     2.39±0.02μs      2.19±0.04μs     0.91  spatial.Xdist.time_pdist(10, 'sokalmichener')
> -     2.46±0.01μs      2.21±0.02μs     0.90  spatial.Xdist.time_pdist(10, 'yule')
> -     2.40±0.02μs      2.15±0.02μs     0.90  spatial.Xdist.time_pdist(10, 'kulsinski')
> -     2.39±0.02μs      2.14±0.02μs     0.89  spatial.Xdist.time_pdist(10, 'dice')
> -     2.94±0.03μs      2.61±0.01μs     0.89  spatial.Xdist.time_cdist(10, 'yule')
> -     2.78±0.04μs      2.46±0.04μs     0.88  spatial.Xdist.time_cdist(10, 'russellrao')
> -     2.43±0.01μs      2.12±0.01μs     0.87  spatial.Xdist.time_pdist(10, 'sokalsneath')
> -     1.14±0.01ms         990±10μs     0.87  spatial.Xdist.time_pdist(1000, 'hamming')
> -     2.41±0.01μs      2.09±0.07μs     0.87  spatial.Xdist.time_pdist(10, 'rogerstanimoto')
> -     2.86±0.02μs      2.47±0.01μs     0.86  spatial.Xdist.time_cdist(10, 'sokalsneath')
> -     2.86±0.01μs      2.46±0.01μs     0.86  spatial.Xdist.time_cdist(10, 'kulsinski')
> -        2.41±0μs      2.07±0.04μs     0.86  spatial.Xdist.time_pdist(10, 'russellrao')
> -     2.83±0.01μs      2.42±0.05μs     0.86  spatial.Xdist.time_cdist(10, 'sokalmichener')
> -      27.3±0.3μs       23.2±0.3μs     0.85  spatial.XdistWeighted.time_cdist(100, 'hamming')
> -     2.89±0.04μs      2.44±0.04μs     0.84  spatial.Xdist.time_cdist(10, 'dice')
> -     2.83±0.02μs      2.38±0.04μs     0.84  spatial.Xdist.time_cdist(10, 'rogerstanimoto')
> -        18.4±1μs      13.1±0.07μs     0.71  spatial.XdistWeighted.time_pdist(100, 'hamming')
> -      26.9±0.1μs      18.6±0.04μs     0.69  spatial.Xdist.time_cdist(100, 'hamming')
> -      16.8±0.1μs      10.5±0.03μs     0.63  spatial.Xdist.time_pdist(100, 'hamming')
> -     7.81±0.02μs      3.57±0.02μs     0.46  spatial.XdistWeighted.time_cdist(20, 'hamming')
> -     7.43±0.09μs      2.99±0.01μs     0.40  spatial.XdistWeighted.time_pdist(20, 'hamming')
> -     7.14±0.09μs      2.84±0.02μs     0.40  spatial.XdistWeighted.time_cdist(10, 'hamming')
> -     7.11±0.08μs      2.66±0.02μs     0.37  spatial.XdistWeighted.time_pdist(10, 'hamming')
> -     6.85±0.06μs      2.37±0.05μs     0.35  spatial.Xdist.time_cdist(10, 'hamming')
> -     6.59±0.09μs      2.08±0.01μs     0.32  spatial.Xdist.time_pdist(10, 'hamming')
> -         251±7μs      2.63±0.06μs     0.01  spatial.XdistWeighted.time_pdist(10, 'russellrao')
> -         335±3μs      2.68±0.01μs     0.01  spatial.XdistWeighted.time_pdist(10, 'jaccard')
> -         510±2μs      2.84±0.03μs     0.01  spatial.XdistWeighted.time_pdist(10, 'yule')
> -         533±1μs      2.91±0.02μs     0.01  spatial.XdistWeighted.time_cdist(10, 'russellrao')
> -       513±0.8μs      2.61±0.02μs     0.01  spatial.XdistWeighted.time_pdist(10, 'sokalmichener')
> -         517±5μs      2.60±0.03μs     0.01  spatial.XdistWeighted.time_pdist(10, 'rogerstanimoto')
> -         540±1μs      2.68±0.03μs     0.00  spatial.XdistWeighted.time_pdist(10, 'kulsinski')
> -         551±5μs      2.66±0.04μs     0.00  spatial.XdistWeighted.time_pdist(10, 'sokalsneath')
> -         740±7μs      2.96±0.03μs     0.00  spatial.XdistWeighted.time_cdist(10, 'jaccard')
> -         688±7μs      2.66±0.01μs     0.00  spatial.XdistWeighted.time_pdist(10, 'dice')
> -     1.04±0.01ms      3.08±0.02μs     0.00  spatial.XdistWeighted.time_pdist(20, 'russellrao')
> -     1.13±0.02ms      3.33±0.09μs     0.00  spatial.XdistWeighted.time_cdist(10, 'yule')
> -     1.16±0.02ms      2.90±0.04μs     0.00  spatial.XdistWeighted.time_cdist(10, 'rogerstanimoto')
> -     1.22±0.01ms      3.02±0.02μs     0.00  spatial.XdistWeighted.time_cdist(10, 'kulsinski')
> -     1.14±0.01ms      2.82±0.01μs     0.00  spatial.XdistWeighted.time_cdist(10, 'sokalmichener')
> -     1.24±0.01ms      3.02±0.08μs     0.00  spatial.XdistWeighted.time_cdist(10, 'sokalsneath')
> -     1.44±0.01ms      3.41±0.05μs     0.00  spatial.XdistWeighted.time_pdist(20, 'jaccard')
> -     1.53±0.01ms      2.98±0.03μs     0.00  spatial.XdistWeighted.time_cdist(10, 'dice')
> -     2.16±0.03ms      3.94±0.04μs     0.00  spatial.XdistWeighted.time_pdist(20, 'yule')
> -     2.16±0.03ms      3.78±0.02μs     0.00  spatial.XdistWeighted.time_cdist(20, 'russellrao')
> -     2.93±0.04ms      4.54±0.03μs     0.00  spatial.XdistWeighted.time_cdist(20, 'jaccard')
> -     2.16±0.03ms      3.14±0.05μs     0.00  spatial.XdistWeighted.time_pdist(20, 'sokalmichener')
> -     2.17±0.01ms      3.13±0.08μs     0.00  spatial.XdistWeighted.time_pdist(20, 'rogerstanimoto')
> -     2.31±0.01ms      3.22±0.02μs     0.00  spatial.XdistWeighted.time_pdist(20, 'kulsinski')
> -     2.31±0.02ms      3.19±0.06μs     0.00  spatial.XdistWeighted.time_pdist(20, 'sokalsneath')
> -     4.55±0.06ms      5.57±0.08μs     0.00  spatial.XdistWeighted.time_cdist(20, 'yule')
> -     2.90±0.02ms      3.36±0.04μs     0.00  spatial.XdistWeighted.time_pdist(20, 'dice')
> -     4.51±0.03ms      3.84±0.08μs     0.00  spatial.XdistWeighted.time_cdist(20, 'rogerstanimoto')
> -     4.73±0.04ms      3.99±0.07μs     0.00  spatial.XdistWeighted.time_cdist(20, 'kulsinski')
> -     4.77±0.03ms      3.98±0.06μs     0.00  spatial.XdistWeighted.time_cdist(20, 'sokalsneath')
> -     4.54±0.05ms      3.75±0.01μs     0.00  spatial.XdistWeighted.time_cdist(20, 'sokalmichener')
> -     6.22±0.07ms      4.34±0.01μs     0.00  spatial.XdistWeighted.time_cdist(20, 'dice')
> -      36.2±0.3ms      24.8±0.03μs     0.00  spatial.XdistWeighted.time_pdist(100, 'jaccard')
> -        55.1±1ms       37.2±0.2μs     0.00  spatial.XdistWeighted.time_pdist(100, 'yule')
> -      72.5±0.1ms       46.7±0.4μs     0.00  spatial.XdistWeighted.time_cdist(100, 'jaccard')
> -         111±1ms       71.0±0.7μs     0.00  spatial.XdistWeighted.time_cdist(100, 'yule')
> -      26.1±0.2ms       16.2±0.2μs     0.00  spatial.XdistWeighted.time_pdist(100, 'russellrao')
> -      52.2±0.3ms       29.4±0.4μs     0.00  spatial.XdistWeighted.time_cdist(100, 'russellrao')
> -      58.9±0.6ms       19.7±0.3μs     0.00  spatial.XdistWeighted.time_pdist(100, 'kulsinski')
> -      73.8±0.3ms       23.7±0.2μs     0.00  spatial.XdistWeighted.time_pdist(100, 'dice')
> -      59.0±0.2ms       18.9±0.1μs     0.00  spatial.XdistWeighted.time_pdist(100, 'sokalsneath')
> -         118±1ms      35.8±0.06μs     0.00  spatial.XdistWeighted.time_cdist(100, 'kulsinski')
> -       149±0.8ms       44.9±0.2μs     0.00  spatial.XdistWeighted.time_cdist(100, 'dice')
> -      55.3±0.7ms       16.3±0.2μs     0.00  spatial.XdistWeighted.time_pdist(100, 'sokalmichener')
> -      55.8±0.4ms       16.4±0.2μs     0.00  spatial.XdistWeighted.time_pdist(100, 'rogerstanimoto')
> -         118±2ms       34.0±0.1μs     0.00  spatial.XdistWeighted.time_cdist(100, 'sokalsneath')
> -       113±0.4ms       29.9±0.3μs     0.00  spatial.XdistWeighted.time_cdist(100, 'rogerstanimoto')
> -       112±0.7ms       29.6±0.3μs     0.00  spatial.XdistWeighted.time_cdist(100, 'sokalmichener')
> 
> SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
> PERFORMANCE DECREASED.
> ```

### CODE REVIEW

**Comment by jjerphan** [comment]
I do not understand why some Cirrus-CI jobs are failing on both this PR and #17522.

**Comment by jjerphan** [comment]
Hi @rgommers, 

I saw you participated in https://github.com/scipy/scipy/pull/14640; who can I ask for reviewing this Pull Request? Should I send a mail on the internal mailing list? :slightly_smiling_face:

File: scipy/spatial/distance.py [hunk 1]
```diff
@@ -1832,16 +1832,16 @@ class MetricInfo:
         types=['double', 'bool'],
         validator=_validate_hamming_kwargs,
         dist_func=hamming,
-        cdist_func=CDistWeightedMetricWrapper('hamming', 'hamming'),
-        pdist_func=PDistWeightedMetricWrapper('hamming', 'hamming'),
+        cdist_func=_distance_pybind.cdist_hamming,
+        pdist_func=PDistMetricWrapper('hamming'),
```
Comment (lorentzenchr) [inline scipy/spatial/distance.py:1836]:
```suggestion
        pdist_func=_distance_pybind.pdist_hamming,
```

**Review by lorentzenchr — APPROVED** [review]
I'm not an expert on those distances but it looks good to me.

**Comment by jjerphan** [comment]
Do we need to add tests for checking the correctness of distance metrics' implementations?

**Comment by jjerphan** [comment]
Gentle ping @peterbell10 and @tylerjereddy: do you have time and are you interested in reviewing this PR? :angel:

**Comment by jjerphan** [comment]
Alternatively, are you interested in reviewing this PR, @Kai-Striega? :slightly_smiling_face:

**Comment by Kai-Striega** [comment]
@jjerphan I'm happy to take a look but won't have time until the weekend

**Comment by jjerphan** [comment]
Thank you! Do not feel in the urge of reviewing it promptly (I don't want to pressure anyone, this PR is not urgent).

**Review by Kai-Striega — APPROVED** [review]
I'm not an expert on the `spatial` module, but, this LGTM. Well done!

> Do we need to add tests for checking the correctness of distance metrics' implementations?

Do you mean testing the C++ functions? I don't believe that is required. Or are there tests missing for the existing Python interfaces? In that case they should be added.

File: scipy/spatial/src/distance_metrics.h [hunk 2]
```diff
@@ -345,3 +345,431 @@ struct CanberraDistance {
...
+        Acc(): nonmatches(0), tt_matches(0) {}
+        T nonmatches, tt_matches;
+    };
+
+    template <typename T>
+    void operator()(StridedView2D<T> out, StridedView2D<const T> x, StridedView2D<const T> y) const {
+        transform_reduce_2d_<2>(out, x, y, [](T x, T y) INLINE_LAMBDA {
+            Acc<T> acc;
+            acc.nonmatches = (x != y);
+            acc.tt_matches = (x != 0) & (y != 0);
```
Comment (peterbell10) [inline scipy/spatial/src/distance_metrics.h:407]: These functions need to be changed slightly to deal with floating point inputs correctly. Basically you need to write everything in terms of logical and which becomes multiplication, and logical not which becomes `1.0 - x`. 
```suggestion
            acc.nonmatches = x * (1.0 - y) + y * (1.0 - x);
            acc.tt_matches = x * y;
```

There is an existing bug in cdist that probably means it isn't tested properly. See #17703

File: scipy/spatial/distance.py [hunk 3]
```diff
@@ -1816,8 +1816,8 @@ class MetricInfo:
         aka={'dice'},
         types=['bool'],
         dist_func=dice,
-        cdist_func=CDistMetricWrapper('dice'),
-        pdist_func=PDistMetricWrapper('dice'),
+        cdist_func=_distance_pybind.cdist_dice,
+        pdist_func=_distance_pybind.pdist_dice,
```
Comment (peterbell10) [inline scipy/spatial/distance.py:1820]: These boolean metrics have gotten slower which I believe is due to promoting their inputs to float. Maybe we could have a wrapper function that dispatches to the old C code for boolean tensors and the new pybind code for everything else?

**Review by peterbell10 — DISMISSED** [review]

**Comment by jjerphan** [comment]
Hi @peterbell10, thank you for your review.

I have two questions:
 - must the old C code be removed?
 - I think splitting this PR is smaller one for each distance metric might make it easier to discuss and review. What do you think?

**Comment by tupui** [comment]
FYI the branching for SciPy 1.11 is planned for May 29th.

**Comment by h-vetinari** [comment]
@peterbell10, can you resolve the open questions by @jjerphan here? Or, if you're not available anymore, remove your request for changes? This PR (incl. its previous iterations) is soon going to be 2 years old.

File: scipy/spatial/distance.py (hunk 3 above)
Comment (h-vetinari) [reply scipy/spatial/distance.py:1820]: Without knowing all the context, I believe a minor performance regression is completely acceptable if it allows us to get rid of the old C code. And looking at the benchmarks in the OP, it's gotten _a lot_ faster elsewhere too, making this even more of a worthwhile trade-off IMO.

File: scipy/spatial/src/distance_metrics.h (hunk 2 above)
Comment (tylerjereddy) [reply scipy/spatial/src/distance_metrics.h]: I rebased, added a regression test for that issue, and the suggested patch that allows the test to pass, and force pushed in. Maybe better now?

File: scipy/spatial/distance.py (hunk 3 above)
Comment (tylerjereddy) [reply scipy/spatial/distance.py:1779]: I took the time to fix the bug, but not sure I'll have the bandwidth to dig into this one, can you live with the this tradeoff @peterbell10?

**Comment by tylerjereddy** [comment]
I'll agree with @h-vetinari on the performance tradeoff concern, but I will open a separate issue for that to track/discuss it with Peter's guidance.

This moves us farther along the C++ migration for the distance metrics and largely improves performance with a few exceptions.

The CI failures are unrelated and I'm pretty happy with the changes here now. As for the removing the old C code, not sure on the timeline, but certainly this could move us closer in that direction for some metrics.

**Pull request merged**